File: backend/internal/service/superadmin_service.go
package service

import (
	"errors"
	"fmt"

	"github.com/ecotrip/backend/internal/model"
	"github.com/ecotrip/backend/internal/repository"
)

type SuperadminStats struct {
	TotalPendingBusinessUsers int64 `json:"total_pending_business_users"`
	TotalApprovedBusiness     int64 `json:"total_approved_business"`
	TotalDestinations         int64 `json:"total_destinations"`
	TotalAccommodations       int64 `json:"total_accommodations"`
}

type SuperadminService interface {
	GetPendingBusinessUsers() ([]model.User, error)
	ApproveBusinessUser(userID uint64) error
	RejectBusinessUser(userID uint64) error
	GetAllDestinations() ([]model.Destination, error)
	GetAllAccommodations() ([]model.Accommodation, error)
	GetStats() (*SuperadminStats, error)
}

type superadminService struct {
	userRepo repository.UserRepository
	destRepo repository.DestinationRepository
	accRepo  repository.AccommodationRepository
}

func NewSuperadminService(
	userRepo repository.UserRepository,
	destRepo repository.DestinationRepository,
	accRepo repository.AccommodationRepository,
) SuperadminService {
	return &superadminService{
		userRepo: userRepo,
		destRepo: destRepo,
		accRepo:  accRepo,
	}
}

func (s *superadminService) GetPendingBusinessUsers() ([]model.User, error) {
	return s.userRepo.FindPendingBusinessUsers()
}

func (s *superadminService) ApproveBusinessUser(userID uint64) error {
	user, err := s.userRepo.FindByID(userID)
	if err != nil {
		return errors.New("pengguna tidak ditemukan")
	}
	if user.Role != "business_destination" && user.Role != "business_accommodation" {
		return errors.New("hanya akun admin pelaku usaha yang memerlukan persetujuan")
	}
	return s.userRepo.UpdateApprovalStatus(userID, "approved")
}

func (s *superadminService) RejectBusinessUser(userID uint64) error {
	user, err := s.userRepo.FindByID(userID)
	if err != nil {
		return errors.New("pengguna tidak ditemukan")
	}
	if user.Role != "business_destination" && user.Role != "business_accommodation" {
		return errors.New("hanya akun admin pelaku usaha yang memerlukan persetujuan")
	}
	return s.userRepo.UpdateApprovalStatus(userID, "rejected")
}

func (s *superadminService) GetAllDestinations() ([]model.Destination, error) {
	return s.destRepo.FindAll()
}

func (s *superadminService) GetAllAccommodations() ([]model.Accommodation, error) {
	return s.accRepo.FindAll()
}

func (s *superadminService) GetStats() (*SuperadminStats, error) {
	pendingCount, approvedBizCount, err := s.userRepo.CountBusinessStats()
	if err != nil {
		return nil, fmt.Errorf("gagal menghitung stats user: %w", err)
	}

	dests, err := s.destRepo.FindAll()
	if err != nil {
		return nil, fmt.Errorf("gagal menghitung stats destinasi: %w", err)
	}

	accs, err := s.accRepo.FindAll()
	if err != nil {
		return nil, fmt.Errorf("gagal menghitung stats akomodasi: %w", err)
	}

	return &SuperadminStats{
		TotalPendingBusinessUsers: pendingCount,
		TotalApprovedBusiness:     approvedBizCount,
		TotalDestinations:         int64(len(dests)),
		TotalAccommodations:       int64(len(accs)),
	}, nil
}

File: frontend/services/superadminService.ts
import api from './api';
import { UserProfile } from './authService';
import { Destination } from './destinationService';
import { Accommodation } from './accommodationService';

export interface SuperadminStats {
  total_pending_business_users: number;
  total_approved_business: number;
  total_destinations: number;
  total_accommodations: number;
}

export const superadminService = {
  async getStats(): Promise<SuperadminStats> {
    const res = await api.get<SuperadminStats>('/superadmin/stats');
    return res.data;
  },

  async getPendingBusinessUsers(): Promise<UserProfile[]> {
    const res = await api.get<{ users: UserProfile[]; count: number }>('/superadmin/business-users/pending');
    return res.data.users || [];
  },

  async approveBusinessUser(userId: number): Promise<{ message: string }> {
    const res = await api.post<{ message: string }>(`/superadmin/business-users/${userId}/approve`);
    return res.data;
  },

  async rejectBusinessUser(userId: number): Promise<{ message: string }> {
    const res = await api.post<{ message: string }>(`/superadmin/business-users/${userId}/reject`);
    return res.data;
  },

  async getAllDestinations(): Promise<Destination[]> {
    const res = await api.get<{ destinations: Destination[]; count: number }>('/superadmin/destinations');
    return res.data.destinations || [];
  },

  async getAllAccommodations(): Promise<Accommodation[]> {
    const res = await api.get<{ accommodations: Accommodation[]; count: number }>('/superadmin/accommodations');
    return res.data.accommodations || [];
  },
};
